azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -36,17 +43,178 @@ def __init__(self, callback):
+        self.completed = False  # Operation has been fully completed
+        self.completing = False  # Operation is in the process of completing
+        self.error = None  # Error associated with Operation completion
+
+        self.add_callback(callback)
+
+    def add_callback(self, callback):
+        """Adds a callback to the Operation that will be triggered upon Operation completion.
+
+        When an Operation is completed, all callbacks will be resolved in LIFO order.
+
+        Callbacks cannot be added to an already completed operation, or an operation that is
+        currently undergoing a completion process.
+
+        :param callback: The callback to add to the operation.
+
+        :raises: OperationError if the operation is already completed, or is in the process of
+            completing.
+        """
+        if self.completed:
+            raise pipeline_exceptions.OperationError(
+                "{}: Attempting to add a callback to an already-completed operation!".format(
+                    self.name
+                )
+            )
+        if self.completing:
+            raise pipeline_exceptions.OperationError(
+                "{}: Attempting to add a callback to a operation with completion in progress!".format(
+                    self.name
+                )
+            )
+        else:
+            self.callback_stack.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def complete(self, error=None):
+        """ Complete the operation, and trigger all callbacks in LIFO order.
+
+        The operation is completed successfully be default, or completed unsucessfully if an error
+        is provided.
+
+        An operation that is already fully completed, or in the process of completion cannot be
+        completed again.
+
+        This process can be halted if a callback for the operation invokes the .halt_completion()
+        method on this Operation.
+
+        :param error: Optionally provide an Exception object indicating the error that caused
+            the completion. Providing an error indicates that the operation was unsucessful.
+        """
+        if error:
+            logger.error("{}: completing with error {}".format(self.name, error))
+        else:
+            logger.debug("{}: completing without error".format(self.name))
+
+        if self.completed or self.completing:
+            logger.error("{}: has already been completed!".format(self.name))
+            e = pipeline_exceptions.OperationError(
+                "Attempting to complete an already-completed operation: {}".format(self.name)
+            )
+            # This could happen in a foreground or background thread, so err on the side of caution
+            # and send it to the background handler.
+            handle_exceptions.handle_background_exception(e)
+        else:
+            # Operation is now in the process of completing
+            self.completing = True
+            self.error = error
```

Comment:
>self.error = error [](start = 12, length = 18)

<grumble> not sure what I think about this.  Now error is a parameter that's passed to the callback and also an attribute of the op.  I think we have to pick one.  If it stays as an attribute, then maybe we should remove it from the callback parameter list.  (not as part of this CR)

Reply:
I agree, this is scheduled for Part 2

---
In reply to: [347139684](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347139684) [](ancestors = 347139684)